### PR TITLE
DOC: Refresh pylibcudf guide

### PR DESCRIPTION
**Pull request by lithomas1**
## Description

This PR updates the pylibcudf dev guide with some more recent recommendations.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by wence-** [comment]
@lithomas1 you'll want to merge branch-24.08 to fix the build failures.

File: docs/cudf/source/developer_guide/pylibcudf.md [hunk 1]
```diff
@@ -242,3 +246,8 @@ cpdef ColumnOrTable empty_like(ColumnOrTable input)
 
 [Cython supports specializing the contents of fused-type functions based on the argument types](https://cython.readthedocs.io/en/latest/src/userguide/fusedtypes.html#type-checking-specializations), so any type-specific logic may be encoded using the appropriate conditionals.
 See the pylibcudf source for examples of how to implement such functions.
+
+In the event that an libcudf overload takes in a different amount of arguments compared to another overload, you should specify the maximum amount of arguments in the Cython definition,
+with the arguments not shared between overloads set to None. If a user tries to pass in an unsupported argument for a specific overload type, you should raise NotImplementedError/ValueError.
+
+Finally, consider making an libcudf issue if you think this inconsistency can be addressed on the libcudf side.
```
[the remark this answers is not in the document]
Comment (lithomas1) [reply docs/cudf/source/developer_guide/pylibcudf.md:253]: I think the libcudf folks have amenable to renaming stuff so far.
(we got the column overload of ``replace`` renamed to ``replace_multiple``)

**Comment by lithomas1** [comment]
> @lithomas1 you'll want to merge branch-24.08 to fix the build failures.

Thanks for the review.

I think I was planning on having this PR sit in draft for a while, until we got most of the pylibcudf API done.
(I've been jotting down most of the stuff as we've finished discussing them.)

I'm definitely happy to address any comments that y'all have in the meantime, though.

**Comment by vyasr** [comment]
Yeah feel free to do this however you see fit. I wasn't planning to review until it was out of draft.

**Review by vyasr — APPROVED** [review]

File: docs/cudf/source/developer_guide/pylibcudf.md [hunk 2]
```diff
@@ -13,10 +13,8 @@ To satisfy the goals of pylibcudf, we impose the following set of design princip
...
-- All cudf code should interact only with pylibcudf, never with libcudf directly.
-- All imports should be relative so that pylibcudf can be easily extracted from cudf later
-  - Exception: All imports of libcudf API bindings in `cudf._lib.cpp` should use absolute imports of `cudf._lib.cpp as libcudf`. We should convert the `cpp` directory into a proper package so that it can be imported as `libcudf` in that fashion. When moving pylibcudf into a separate package, it will be renamed to `libcudf` and only the imports will need to change.
-- Ideally, pylibcudf should depend on nothing other than rmm and pyarrow. This will allow it to be extracted into a a largely standalone library and used in environments where the larger dependency tree of cudf may be cumbersome.
+- All cudf code should interact only with pylibcudf, never with libcudf directly. This is not currently the case, but is the direction that the library is moving towards.
+- Ideally, pylibcudf should depend on nothing other than rmm. This will allow it to be extracted into a a largely standalone library and used in environments where the larger dependency tree of cudf may be cumbersome.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/pylibcudf.md:17]: "will allow", this has been done, no?

**Review by wence- — APPROVED** [review]

File: docs/cudf/source/developer_guide/pylibcudf.md (hunk 2 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/pylibcudf.md:17]: Indeed, thanks.

Comment (vyasr) [inline docs/cudf/source/developer_guide/pylibcudf.md:17]:
```suggestion
- Ideally, pylibcudf should depend on no RAPIDS component other than rmm, and should in general have minimal runtime dependencies.
```

**Pull request merged**